py/bitbox02/bitbox02/communication/bitbox_api_protocol.py

```diff
@@ -270,21 +301,133 @@ def _create_noise_channel(self, noise_config: BitBoxNoiseConfig) -> NoiseConnect
+    """ BitBox Protocol from firmware V2.0.0 onwards. """
+
+    def unlock_query(self) -> bytes:
+        unlock_result, unlock_data = self.query(OP_UNLOCK, b"")
+        if len(unlock_data) != 0:
+            raise ValueError(f"OP_UNLOCK (V3) replied with wrong length.")
+        return unlock_result
+
+
+class BitBoxProtocolV3(BitBoxProtocolV2):
+    """ BitBox Protocol from firmware V3.0.0 onwards. """
+
+    def unlock_query(self) -> bytes:
+        unlock_result, unlock_data = self.query(OP_UNLOCK, b"")
+        if len(unlock_data) != 0:
+            raise ValueError(f"OP_UNLOCK (V3) replied with wrong length.")
+        return unlock_result
+
+
+class BitBoxProtocolV4(BitBoxProtocolV3):
+    """ BitBox Protocol from firmware V4.0.0 onwards. """
+
+    def _encode_noise_request(self, encrypted_msg: bytes) -> bytes:
+        return OP_NOISE_MSG + encrypted_msg
+
+
+class BitBoxProtocolV7(BitBoxProtocolV4):
+    """ Noise Protocol from firmware V7.0.0 onwards. """
+
+    def _handshake_query(self, req: bytes) -> Tuple[bytes, bytes]:
+        return self.query(OP_HER_COMEZ_TEH_HANDSHAEK, req)
+
+    def _decode_noise_response(self, encrypted_msg: bytes) -> Tuple[bytes, bytes]:
+        print("Decoding: " + encrypted_msg.hex())
+        return encrypted_msg[0:1], encrypted_msg[1:]
+
+
+class BitBoxCommonAPI:
+    """Class to communicate with a BitBox device"""
+
+    # pylint: disable=too-many-public-methods,too-many-arguments
+    def __init__(
+        self, transport: TransportLayer, device_info: DeviceInfo, noise_config: BitBoxNoiseConfig
+    ):
+        self.debug = False
+        serial_number = device_info["serial_number"]
+
+        self.version = parse_device_version(serial_number)
+        if self.version is None:
+            transport.close()
+            raise ValueError(f"Could not parse version from {serial_number}")
+        # Delete the prelease part, as it messes with the comparison (e.g. 3.0.0-pre < 3.0.0 is
+        # True, but the 3.0.0-pre has already the same API breaking changes like 3.0.0...).
+        self.version = semver.VersionInfo(
+            self.version.major, self.version.minor, self.version.patch, build=self.version.build
+        )
+
+        self._bitbox_protocol: BitBoxProtocol
+        if self.version >= semver.VersionInfo(7, 0, 0):
+            self._bitbox_protocol = BitBoxProtocolV7(transport)
+        elif self.version >= semver.VersionInfo(4, 0, 0):
+            self._bitbox_protocol = BitBoxProtocolV4(transport)
+        elif self.version >= semver.VersionInfo(3, 0, 0):
+            self._bitbox_protocol = BitBoxProtocolV3(transport)
+        else:
+            self._bitbox_protocol = BitBoxProtocolV1(transport)
+
+        if self.version >= semver.VersionInfo(2, 0, 0):
+            noise_config.attestation_check(self._perform_attestation())
+            unlock_result = self._bitbox_protocol.unlock_query()
+            if unlock_result != RESPONSE_SUCCESS:
```

Comment:
This is only for version 3.0.0+ according to the previous code. in 2.0.0, the response was empty (not even a status code, see https://github.com/digitalbitbox/bitbox02-firmware/commit/40ea9cb4811fe041413da52234822f657c7c28e8#diff-fb3b76656f7207bf1004582fd62acadaR85

Before this PR this was handled explicitly like this:

```Python
            unlock_result = self._query(OP_UNLOCK)
            if self.version < semver.VersionInfo(3, 0, 0):
                assert unlock_result == b""
            else:
                # since 3.0.0, unlock can fail if cancelled
                if unlock_result == RESPONSE_FAILURE:
                    self.close()
                    raise Exception("Unlock process aborted")
```

The pattern of always returning a status code does not work for all previous versions.

Unless I am reading the code wrong, unlock_result for v2.0.0 will be `b""`, so it will always fail for v2.0.0?

Reply:
Otherwise I tested all upgrades from v3.0.0 onwards and those worked. 2.0.0 is granted a low prio as it was a pre-release version, but it's probably good to keep it right anyway.

Reply:
Fixed in #362